### DISCUSSION TITLE
Is there a global events API within NextJS?

### DISCUSSION THREAD

**Discussion by ataschz** [vercel/next.js#58705] [Help]
### Summary

I'm thinking of conducting some experiments with NextJS, and I would like to have a dashboard or app that allows me to easily view and inspect the requests made in my NextJS application, both server-side and client-side.

I've been exploring the NextJS source code, and I believe there isn't something exactly like what I need. However, the information I require does exist; it's just not being stored or utilized. If we use the logging option in `next.config.js` with `fullUrl: true`, we can see API calls from server actions in the console.

On the other hand, by inspecting the code in the `next-server.ts` file, I managed to identify the places where the server sends updated resources to the client. It would be great to be able to intercept these chunks/changes, manipulate them, and display a visual timeline of our application to control caching and the rendering process.

Any ideas on where I should focus my efforts, or if there's any undocumented API that allows access to this information?

**Comment by icyJoseph** [comment]
Hi, 

If you just care about network requests, you can use `@mswjs/interceptors`:

```ts
// app/interceptors.ts
import { BatchInterceptor } from "@mswjs/interceptors";
import { ClientRequestInterceptor } from "@mswjs/interceptors/ClientRequest";
import { XMLHttpRequestInterceptor } from "@mswjs/interceptors/XMLHttpRequest";
import { FetchInterceptor } from "@mswjs/interceptors/fetch";

const interceptor = new BatchInterceptor({
  name: "my-interceptor",
  interceptors: [
    new ClientRequestInterceptor(),
    new XMLHttpRequestInterceptor(),
    new FetchInterceptor(),
  ],
});

interceptor.apply();

// This "request" listener will be called on both
// "http.ClientRequest" and "XMLHttpRequest" being dispatched.
interceptor.on("request", console.log);
interceptor.on("response", console.log);
```

And then use it like:

```ts
// app/layout.tsx
import "./interceptors";
import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";

// rest of layout file
```

IIRC, that'll give you a view of everything that happens. I used that to hack together a little CLI that showed requests and such :)

**Reply by ataschz** [reply]
Thanks! 

This only works to inspect the client-side requests, I'm looking for some wlse to inspect and interact it with the requests "server-side", for example:

I have a simple app with server-side pagination to show a list of items, when I click into "next", the server request a new page and rehydrate the page with the new data sending a new RSC, so, I want to "inspect" and "interact" with this chunks/streams/requests.

With your code @icyJoseph I can see the first request (when I enter to the page), but, then, I can't see the requests but on the console I can see the requests "server-side"

![image](https://github.com/vercel/next.js/assets/9469502/740193ec-498d-4771-b387-38170640b36f)
